### PR TITLE
profiles: support transcode for empty profiles and bitrate-only profiles

### PR DESCRIPTION
**Pull request by emranemran**
The transcode profile logic is being updated to handle these new changes:
* The transcode-api will now pass through empty profiles to catalyst-api. These cases should use catalyst-api's default ABR ladder.
* The transcode-api will also allow profiles with just the bitrate field set. These cases should use the resolution of the input video with the specified target bitrate for transcoding.

TODO: add/fix tests

### CODE REVIEW

File: video/profiles.go [hunk 1]
```diff
@@ -85,11 +85,44 @@ var DefaultProfile720p = EncodedProfile{
...
 	}
+	// If Profiles haven't been overridden, use the default set
+	//get
+	if len(reqTranscodeProfiles) == 0 {
+		transcodeProfiles, err = GetDefaultPlaybackProfiles(videoTrack)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get default transcode profiles: %w", err)
+		}
+		return transcodeProfiles, nil
+	} else if len(reqTranscodeProfiles) == 1 {
```
Comment (thomshutt) [inline video/profiles.go:103]: Let's get a comment in here about why 1 profile is a special case

File: handlers/schemas/UploadVOD.yaml [hunk 2]
```diff
@@ -110,9 +110,6 @@ properties:
       additionalProperties: false
       required:
       -  "name"
-      - "width"
-      -  "height"
-      - "bitrate"
```
[the remark this answers is not in the document]
Comment (emranemran) [reply handlers/schemas/UploadVOD.yaml:115]: Added new commit to check including test case.

File: video/profiles.go [hunk 3]
```diff
@@ -85,11 +85,44 @@ var DefaultProfile720p = EncodedProfile{
...
+		if reqTranscodeProfiles[0].Width == 0 && reqTranscodeProfiles[0].Height == 0 && reqTranscodeProfiles[0].Bitrate != 0 {
+			transcodeProfiles = GetBitrateMatchedProfile(videoTrack, reqTranscodeProfiles[0].Bitrate)
+			return transcodeProfiles, nil
+		}
+	}
+	return reqTranscodeProfiles, nil
+}
+
+func GetBitrateMatchedProfile(videoTrack InputTrack, videoBitrate int64) []EncodedProfile {
```
[the remark this answers is not in the document]
Comment (emranemran) [reply video/profiles.go]: Done

File: video/profiles.go (hunk 1 above)
Comment (emranemran) [reply video/profiles.go:104]: Done

File: handlers/upload.go [hunk 4]
```diff
@@ -73,6 +73,35 @@ func HasContentType(r *http.Request, mimetype string) bool {
...
 
+func (r UploadVODRequest) IsProfileValid() bool {
+	// an empty profile is valid and tells us to use the default ABR ladder
+	if len(r.Profiles) == 0 {
+		return true
+	}
+
+	// a special case where only the bitrate is set which tells us to
+	// generate a profile that matches the input video's specs with the
+	// user specified target bitrate
```
Comment (thomshutt) [inline handlers/upload.go:84]: 👌 glorious

File: handlers/upload.go [hunk 5]
```diff
@@ -73,6 +73,35 @@ func HasContentType(r *http.Request, mimetype string) bool {
...
+	// verify that width/height/bitrate is provided in cases where the
+	// user wants to use their own transcode profile
+	for _, profile := range r.Profiles {
+		if profile.Width == 0 && profile.Height != 0 {
+			return false
+		} else if profile.Width != 0 && profile.Height == 0 {
+			return false
+		} else if profile.Width != 0 && profile.Height != 0 && profile.Bitrate == 0 {
+			return false
+		}
```
Comment (thomshutt) [inline handlers/upload.go:99]: Could this be simplified to `profile.Width == 0 || profile.Height == 0 || profile.Bitrate == 0`?

**Review by thomshutt — APPROVED** [review]

File: handlers/upload.go (hunk 5 above)
Comment (emranemran) [reply handlers/upload.go]: Yeah, I had a few other checks there that got removed. Simplified it further.

**Pull request merged**